src/server/failed_logins.rs

```diff
@@ -0,0 +1,188 @@
+use crate::options::FailedLoginsPenalty;
+use crate::options::FailedLoginsPolicy;
+
+use super::shutdown;
+use slog::Logger;
+use std::collections::HashMap;
+use std::net::IpAddr;
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+use tokio::sync::{Mutex, RwLock};
+
+#[derive(Hash, Eq, PartialEq, Debug, Clone)]
+struct FailedLoginsKey {
+    ip: Option<IpAddr>,
+    username: Option<String>,
+}
+
+#[derive(Debug, Clone)]
+struct FailedLoginsEntry {
+    attempts: u32,
+    last_attempt_at: Instant,
+}
+
+impl FailedLoginsEntry {
+    fn new() -> Mutex<FailedLoginsEntry> {
+        Mutex::new(FailedLoginsEntry {
+            attempts: 1,
+            last_attempt_at: Instant::now(),
+        })
+    }
+
+    fn time_elapsed(&self) -> Duration {
+        self.last_attempt_at.elapsed()
+    }
+
+    fn touch(&mut self) {
+        self.last_attempt_at = Instant::now();
+    }
+}
+
+/// Temporarily remembers failed logins
+#[derive(Debug)]
+pub struct FailedLoginsCache {
+    policy: FailedLoginsPolicy,
+    penalty: FailedLoginsPenalty,
+    failed_logins: Arc<RwLock<HashMap<FailedLoginsKey, Mutex<FailedLoginsEntry>>>>,
+}
+
+#[derive(Debug)]
+pub enum FailedLoginsError {
+    // With this failed login attempt the lockout threshold has been reached
+    MaxFailuresReached,
+    // The account is already locked out from previous failed attempts
+    AlreadyLocked,
+}
+
+impl FailedLoginsCache {
+    pub fn new(failed_logins_policy: FailedLoginsPolicy) -> Arc<FailedLoginsCache> {
```

Comment:
suggest `policy: FailedLoginsPolicy` to reduce the repetition and make for less to read :-)